### PR TITLE
Ersi lig 3912 refactor mae to use timm vit

### PR DESCRIPTION
**Pull request by ersi-lightly**

### CODE REVIEW

File: lightly/models/modules/masked_autoencoder_timm.py [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+            input:
+               Input tensor with shape (batch_size, num_sequences).
+
+        """
+        # code copied from:
+        # https://github.com/facebookresearch/msn/blob/4388dc1eadbe3042b85d3296d41b9b207656e043/src/deit.py#L291
+        npatch = input.shape[1] - 1
+        N = self.pos_embed.shape[1] - 1
+        if npatch == N:
+            return torch.Tensor(self.pos_embed)
```
Comment (guarin) [inline lightly/models/modules/masked_autoencoder_timm.py:256]: No need for new tensor

File: lightly/models/modules/masked_autoencoder_timm.py [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+        decoder_num_heads: int = 16,
+        mlp_ratio: float = 4.0,
+        proj_drop_rate: float = 0.0,
+        attn_drop_rate: float = 0.0,
+        norm_layer: Callable[..., nn.Module] = partial(LayerNorm, eps=1e-6),
+    ):
+        super().__init__()
+
+        self.decoder_embed = nn.Linear(embed_dim, decoder_embed_dim, bias=True)
+        self.mask_token = nn.Parameter(torch.zeros(1, 1, decoder_embed_dim))
```
Comment (guarin) [inline lightly/models/modules/masked_autoencoder_timm.py:355]: We have to decide where the mask token is stored. Right now it is here but there is also a mask token in `examples/pytorch/mae_timm.py` in the main module. From the example code I think the one here is never used (except in the initialization).

File: examples/pytorch/mae_timm.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+class MAE(nn.Module):
+    def __init__(self, vit):
+        super().__init__()
+
+        decoder_dim = 512
+        self.mask_ratio = 0.75
+        self.patch_size = vit.patch_embed.patch_size[0]
+        self.sequence_length = vit.patch_embed.num_patches + 1
+        self.mask_token = nn.Parameter(torch.zeros(1, 1, decoder_dim))
```
Comment (guarin) [inline examples/pytorch/mae_timm.py:29]: Here is the other mask token :)

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks, this looks really promising! I left a few comments :)

File: lightly/models/modules/masked_autoencoder_timm.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        if idx_keep is not None:
+            input = utils.get_at_index(input, idx_keep)
+        # apply Transformer blocks
+        for blk in self.blocks:
+            input = blk(input)
+        # normalize
+        out = self.norm(input)
+        return torch.Tensor(out)
+
+    def interpolate_pos_encoding(self, input: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (ersi-lightly) [reply lightly/models/modules/masked_autoencoder_timm.py]: Nice catch, thanks :) It is in fact controlled by `dynamic_img_size` .

File: lightly/models/modules/masked_autoencoder_timm.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+
+        Args:
+            images:
+                Tensor with shape (batch_size, channels, image_size, image_size).
+
+        Returns:
+            Tensor with shape (batch_size, sequence_length - 1, hidden_dim)
+            containing the patch tokens.
+        """
+        tokens = torch.Tensor(self.patch_embed(images))
```
[the remark this answers is not in the document]
Comment (ersi-lightly) [reply lightly/models/modules/masked_autoencoder_timm.py]: The return type of the `forward` of `self.patch_embed` is `Any`. Thanks for the suggestion!

File: lightly/models/modules/masked_autoencoder_timm.py [hunk 6]
```diff
@@ -60,6 +59,8 @@ class MAEBackbone(vision_transformer.VisionTransformer):  # type: ignore
             Number of register tokens.
         fc_norm:
             Pre head norm after pool (instead of before), if None, enabled when global_pool == 'avg'.
+        dynamic_img_size:
+            If set to True, encoding of variable sized images is handled.
```
Comment (guarin) [inline lightly/models/modules/masked_autoencoder_timm.py:63]: Could you also add the docstring for `dynamic_image_pad`?

File: lightly/models/modules/masked_autoencoder_timm.py [hunk 7]
```diff
@@ -226,46 +244,15 @@ def encode(
         # convert images to tokens and add class token if needed
         input = self.images_to_tokens(images, prepend_class_token=True)
         # add positional encoding
-        input = input + self.interpolate_pos_encoding(input)
+        input = input + self.pos_embed
```
Comment (guarin) [inline lightly/models/modules/masked_autoencoder_timm.py:247]:
```suggestion
        input = self._pos_embed(input)
```

This should also take care of the interpolation for the positional embeddings. It calls `input + self.pos_embed` internally. See: https://github.com/huggingface/pytorch-image-models/blob/992976f007a1f81a1653686ce90be917f50fc0f8/timm/models/vision_transformer.py#L586

This should be a bit less hacky than the way the MAE codebase does it.

**Review by guarin — COMMENTED** [review]
Thanks for the changes! This looks good :) Left some very small comments. Once benchmarks run through we should be ready to merge.

File: lightly/models/modules/masked_autoencoder_timm.py (hunk 7 above)
Comment (ersi-lightly) [reply lightly/models/modules/masked_autoencoder_timm.py:247]: I am not 100% convinced by this suggestion TBH. Just because the `_pos_embed` is a private method of `VisionTransformer`. WDYT?

Comment (ersi-lightly) [reply lightly/models/modules/masked_autoencoder_timm.py]: Removed also the `images_to_tokens` method since the addition of the classification token is also handled by `VisionTransformer`.

File: docs/source/getting_started/install.rst [hunk 8]
```diff
@@ -32,6 +32,13 @@ If you want to work with video files you need to additionally install
 
     pip install av
 
+If you want to work use the Masked Autoencoder you need to additionally install
+`TIMM <https://timm.fast.ai/>`_.
```
Comment (guarin) [inline docs/source/getting_started/install.rst:36]: This link should point to https://github.com/huggingface/pytorch-image-models Timm is not part of fastai.

File: examples/pytorch/mae.py [hunk 9]
```diff
@@ -17,31 +28,33 @@ def __init__(self, vit):
...
-            out_dim=vit.patch_size**2 * 3,
-            dropout=0,
-            attention_dropout=0,
+        self.patch_size = vit.patch_embed.patch_size[0]
+
+        self.backbone = masked_vision_transformer_timm.MaskedVisionTransformerTIMM(
+            vit=vit
+        )
+        self.sequence_length = self.backbone.sequence_length
+        self.decoder = masked_autoencoder_timm.MAEDecoder(
```
Comment (guarin) [inline examples/pytorch/mae.py:37]:
```suggestion
        self.backbone = MaskedVisionTransformerTIMM(
            vit=vit
        )
        self.sequence_length = self.backbone.sequence_length
        self.decoder = MAEDecoder(
```
Let's import the classes directly.

Please also change in the other files.

File: examples/pytorch_lightning/msn.py [hunk 10]
```diff
@@ -21,14 +23,16 @@ def __init__(self):
 
         # ViT small configuration (ViT-S/16)
         self.mask_ratio = 0.15
-        self.backbone = MAEBackbone(
+        # ViT small configuration (ViT-S/16)
+        vit = torchvision.models.VisionTransformer(
```
Comment (guarin) [inline examples/pytorch_lightning/msn.py:27]:
```suggestion
        vit = VisionTransformer(
```

File: examples/pytorch_lightning_distributed/pmsn.py [hunk 11]
```diff
@@ -115,6 +118,7 @@ def configure_optimizers(self):
     devices="auto",
     accelerator="gpu",
     strategy="ddp",
-    use_distributed_sampler=True,  # or replace_sampler_ddp=True for PyTorch Lightning <2.0
+    # use_distributed_sampler=True,  # or replace_sampler_ddp=True for PyTorch Lightning <2.0
+    replace_sampler_ddp=True,
```
Comment (guarin) [inline examples/pytorch_lightning_distributed/pmsn.py:122]: Please revert or change it in all distributed training examples.

File: lightly/models/modules/__init__.py [hunk 12]
```diff
@@ -31,15 +31,24 @@
...
         MAEEncoder,
     )
+    from lightly.models.modules.masked_vision_transformer_torchvision import (
+        MaskedVisionTransformerTorchvision,
+    )
+
 if _dependency.timm_vit_available():
     # Requires timm >= 0.9.9
     from lightly.models.modules.heads_timm import AIMPredictionHead
+    from lightly.models.modules.masked_autoencoder_timm import MAEDecoder
```
Comment (guarin) [inline lightly/models/modules/__init__.py:48]: This doesn't work because this import now overwrites the `MAEDecoder` import from `lightly.models.modules.masked_autoencoder`.

File: lightly/models/modules/__init__.py [hunk 13]
```diff
@@ -31,15 +31,24 @@
 from lightly.utils import dependency as _dependency
 
 if _dependency.torchvision_vit_available():
+    # Requires torchvision >=0.12
```
Comment (guarin) [inline lightly/models/modules/__init__.py:34]: Duplicate comment

**Review by guarin — CHANGES_REQUESTED** [review]
Thanks! Left some comments :)

File: lightly/models/modules/masked_autoencoder_timm.py [hunk 14]
```diff
@@ -1,15 +1,13 @@
...
 from functools import partial
-from typing import Callable, Literal, Optional, Tuple, Type, Union
+from typing import Callable, Optional
 
-try:
-    from timm.layers import LayerType, Mlp, PatchEmbed
+from lightly.utils import dependency
+
+if dependency.timm_vit_available():
     from timm.models import vision_transformer
```
Comment (guarin) [inline lightly/models/modules/masked_autoencoder_timm.py:9]: No need to add the check here. The import of the file will anyways fail if timm is not available.

File: lightly/models/modules/__init__.py (hunk 12 above)
Comment (ersi-lightly) [reply lightly/models/modules/__init__.py:48]: In `masked_autoencoder` we use Torchvision, whereas in `masked_autoencoder_timm` we use TIMM

File: tests/models/modules/test_masked_autoencoder.py [hunk 15]
```diff
@@ -87,16 +193,6 @@ def test_forward(self):
     def test_forward_cuda(self):
         self._test_forward(torch.device("cuda"))
 
-    def test_images_to_tokens(self) -> None:
```
[the remark this answers is not in the document]
Comment (ersi-lightly) [reply tests/models/modules/test_masked_autoencoder.py]: by accident - reverted

File: tests/models/modules/test_masked_autoencoder_timm.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+from lightly.models import utils
+from lightly.utils import dependency
+
+if dependency.timm_vit_available():
+    from timm.models.vision_transformer import VisionTransformer, vit_base_patch32_224
+
+    from lightly.models.modules import MAEDecoderTIMM, MaskedVisionTransformerTIMM
+
+    class TestMAEBackboneTIMM(unittest.TestCase):
```
Comment (guarin) [inline tests/models/modules/test_masked_autoencoder_timm.py:13]: Let's fix this in the next PR: This should be `TestMaskedVisionTransformerTIMM` and in `test_masked_vision_transformer_timm.py`

File: tests/models/modules/test_masked_autoencoder.py [hunk 17]
```diff
@@ -1,12 +1,13 @@
 import unittest
 
 import torch
-import torchvision
 
 from lightly.models import utils
 from lightly.utils import dependency
 
 if dependency.torchvision_vit_available():
+    import torchvision
```
Comment (guarin) [inline tests/models/modules/test_masked_autoencoder.py:9]: Moving the torchvision import inside the `if` is not really required. torchvision should always be installed.

File: tests/models/modules/test_masked_autoencoder_timm.py (hunk 16 above)
Comment (ersi-lightly) [reply tests/models/modules/test_masked_autoencoder_timm.py:13]: hm but we have also the tests for the decoder with TIMM. I can rename the class to `TestMaskedVisionTransformerTIMM `, but I would not change the name of the file TBH

File: tests/models/modules/test_masked_autoencoder.py (hunk 17 above)
Comment (guarin) [reply tests/models/modules/test_masked_autoencoder.py:9]: `torchvision_vit_available` is only `True` if the installed torchvision version includes the VisionTransformer (was added in pytorch 2.0 if I am not mistaken). Torchvision is always installed as per our requirements: https://github.com/lightly-ai/lightly/blob/master/requirements/torch.txt

**Comment by ersi-lightly** [comment]
Closed in favour of #1510

File: tests/models/modules/test_masked_autoencoder_timm.py (hunk 16 above)
Comment (guarin) [reply tests/models/modules/test_masked_autoencoder_timm.py:13]: Please move it to a new file. If you write tests for functions/classes from `dir/something.py` then the corresponding tests should go to `tests/dir/test_something.py`

Comment (ersi-lightly) [reply tests/models/modules/test_masked_autoencoder_timm.py:13]: That is what I have. The test for `lightly/models/modules/masked_autoencoder_timm.py` is in `tests/models/modules/test_masked_autoencoder_timm.py`
